### DISCUSSION TITLE
<items label="%tree.schema.node.name" path="schema" property="schemas"  Some of the problems

### DISCUSSION THREAD

**Discussion by yangmeng66** [dbeaver/dbeaver#15125] [Q&A]
What's the difference between this "property" and "path"? And how to "org. Jkiss. Dbeaver. Ext. Oracle model. The OracleSchema" relates the attributes of a class? Thank you very much

```
<folder type="org.jkiss.dbeaver.ext.oracle.model.OracleSchema" label="%tree.schemas.node.name" icon="#folder_schema" description="Database schemas">
                    <items label="%tree.schema.node.name" path="schema" property="schemas" icon="#schema">
                        <icon if="object.system" icon="#schema_system"/>
.......
```

**Comment by LonwoLonwo** [comment]
Hello @yangmeng66 

Please tell us - what specific goal do you want to achieve?

**Reply by yangmeng66** [reply]
Hello, @ LonwoLonwo!

Well, my company is currently developing our own database (similar to Oracle) and my overall goal is to integrate it into the DBeaver plug-in like Oracle, but there is no documentation on plug-in development and so I have to read the source code of the Oracle plug-in and understand it bit by bit

**Reply by ShadelessFox** [reply]
Perhaps we should write an article about general aspects of plugin development.

There have been a lot of contributions and overall interest recently in DBeaver.

**Comment by ShadelessFox** [comment]
Hello @yangmeng66,

Those are very different properties. Here's a brief description of each:
- `property` defines a _getter_ method used to retrieve children for this folder
- `path` defines a unique name for the navigator node used by the navigator itself and editors (e.g., restoring editors after restart)

The `type` attribute of a `folder` node defines the class the children of this folder are instances of. It serves several purposes:
- The navigator uses it to retrieve all the nested folders of this node to display in its tree
- The metadata editor uses it to display all the properties of this node (and also nested folders at the bottom, just like the navigator does)

Properties are defined in the source file rather than in the `.xml` file. To become property, a _getter_ method must be annotated with the `@Property` attribute. Those are collected via Reflection API from the specified class.

I hope this explanation will help you!

**Reply by yangmeng66** [reply]
Hello, @ shadelessfox!

First of all, thank you for your answer. As you said, < property > is indeed the final getter method (I can only understand it bit by bit). How do you know the meaning of these configurations? my company is currently developing our own database (similar to Oracle) and my overall goal is to integrate it into the DBeaver plug-in like Oracle

**Reply by ShadelessFox** [reply]
> How do you know the meaning of these configurations?

I'm a bit confused. What are you talking about exactly?

I work on DBeaver every day, so I'm familiar with all that stuff if this is what you were asking about.

**Reply by yangmeng66** [reply]
Sorry, my English is not very good. What I want to say is: is there a reference document for the configuration of plug-in development (including code development manual)? Now I can only look at the source of Oracle plug-in to understand it bit by bit, because I finally want to integrate our company's database into dbeaver, just like oracle

**Reply by ShadelessFox** [reply]
Unfortunately, there are no docs or references about plugin development.

So the only way is to use the existing code as the reference.

**Reply by yangmeng66** [reply]
Oh, I see. I'll look at the existing Oracle code to understand it. If I don't understand anything, I'll consult you. Thank you very much

**Reply by yangmeng66** [reply]
Hello, @ShadelessFox:
 I have basically understood the configuration < extensionpoint = "org. Jkiss. Dbeaver. Datasourceprovider" >
, other related < extension point = "..." > If you have spare time, please guide me to the general meaning of this configuration so that I can understand it faster and more accurately. Thank you very much



    <extension point="org.jkiss.dbeaver.objectManager">      
    <extension point="org.jkiss.dbeaver.dataTypeProvider">   
    <extension point="org.jkiss.dbeaver.dataSourceAuth">     
    <extension point="org.eclipse.core.runtime.adapters">   
    <extension point="org.jkiss.dbeaver.task">              
    <extension point="org.eclipse.ui.bindings">             
    <extension point="org.eclipse.core.runtime.preferences"> 
    <extension point="org.jkiss.dbeaver.sqlInsertMethod">    
    <extension point="org.jkiss.dbeaver.sqlDialect">

**Reply by ShadelessFox** [reply]
In particular, which parts are you interested in?

**Reply by yangmeng66** [reply]
Hello, I'm very sorry. Maybe I asked too many questions. I don't know what the functions of many parts are, and I naturally can't choose. I hope you can help me explain the general meaning of these < extension > (don't be too detailed)

```
    <extension point="org.jkiss.dbeaver.objectManager">      --incomprehension
    <extension point="org.jkiss.dbeaver.dataTypeProvider">    --Personal understanding: Add new data type of database
    <extension point="org.jkiss.dbeaver.dataSourceAuth">     --incomprehension
    <extension point="org.eclipse.core.runtime.adapters">     --incomprehension
    <extension point="org.jkiss.dbeaver.task">               --incomprehension
    <extension point="org.eclipse.ui.bindings">             --incomprehension
    <extension point="org.eclipse.core.runtime.preferences">      --incomprehension
    <extension point="org.jkiss.dbeaver.sqlInsertMethod">       --incomprehension
    <extension point="org.jkiss.dbeaver.sqlDialect">          ---Personal understanding: Add  Unique syntax keywords  of database
```

Other, what is the function of this module? How are the two modules connected


![image](https://user-images.githubusercontent.com/51857567/148735331-8a1f8d95-d850-42b8-a797-11142faf37d9.png)

**Reply by ShadelessFox** [reply]
This module implements the UI part of the extension.

We separate model and UI modules so that we can use the standalone model in CloudBeaver, or to run DBeaver headless.

**Reply by ShadelessFox** [reply]
Here's a really brief description of each:
 - `org.jkiss.dbeaver.objectManager` defines a manager for a given model object. It provides actions such as object creation, deletion, editing, etc.
 - `org.jkiss.dbeaver.dataSourceAuth` defines custom authorization methods for a data source.
 - `org.jkiss.dbeaver.task` defines new task types used by the [Tasks](https://github.com/dbeaver/dbeaver/wiki/Task-Management).
 - `org.jkiss.dbeaver.sqlInsertMethod` defines a set of supported `INSERT IGNORE` variations by the data source.

For `org.eclipse.*` extension points, please look at Eclipse documentation.

**Reply by yangmeng66** [reply]
Oh, I understand a lot, but one more thing is: why customize this .....auth, I don't know much about application scenarios

extension point="org.jkiss.dbeaver.dataSourceAuth"

**Reply by ShadelessFox** [reply]
A database may support IAM/Kerberos etc. authorization schemes.

DBeaver knows nothing about them. By creating an extension point with the proper implementation, we can provide support for such authorization schemes for a data source.

**Reply by yangmeng66** [reply]
Oh, one more thing, I didn't understand what you said?    I found that some modules do not have ".... UI", such as Clickhouse

![image](https://user-images.githubusercontent.com/51857567/148746923-9a264435-527f-495d-bc70-8e46ee5fa9a7.png)

**Reply by ShadelessFox** [reply]
The reason why we have both `org.jkiss.dbeaver.ext.oracle` and `org.jkiss.dbeaver.ext.oracle.ui` is that because we want to separate model logic from UI logic. It allows us to run DBeaver without any UI (headless), or reuse the existing model in CloudBeaver.

It is not necessary to have a UI bundle. It is only required when you want to provide a specific UI for your model. An example would be the role editor in the PostgreSQL bundle. You can take a look at `org.jkiss.dbeaver.ext.postgresql.ui.editors.PostgresRolePrivilegesEditor`.

Note: _bundle_ is the same as _module_.

**Reply by yangmeng66** [reply]
Like this,  need to customize the special interface,        This is when the "... UI" is needed

![image](https://user-images.githubusercontent.com/51857567/148751405-3e5e4025-48a4-4c87-93cd-fa63648b2fa3.png)

**Reply by ShadelessFox** [reply]
Yes, exactly.

**Reply by yangmeng66** [reply]
hi, @ShadelessFox 
The default universal JDBC is displayed by default after connecting to dbeaver, such as Clickhouse. Oracle is a user-defined interface. My question is: if I want to implement the Oracle interface, should I completely customize all hierarchical queries instead of using the default display of dbeaver???

**Reply by ShadelessFox** [reply]
I'm not sure which interface are you referring to.

In the case of the navigator tree, Clickhouse uses a generic driver that has a predefined tree structure. In contrast, Oracle is not a generic driver, so it has to define its own tree structure.

**Reply by yangmeng66** [reply]
hi
Like this, the Clickhouse has a default generic cdriver tree structure, but now I want to put the default "default, system" in a custom "Schemas" directory. In this case, can I not use "Clickhouse uses a generic driver that has a predefined tree structure"???? 

![image](https://user-images.githubusercontent.com/51857567/148915356-6066394e-0dfd-42ea-9e5d-618e9aaff86b.png)

Do I need to customize it all?  Like Oracle
![image](https://user-images.githubusercontent.com/51857567/148918048-2ecd1145-9d3b-4f80-a1f3-10e6463724d2.png)

**Reply by ShadelessFox** [reply]
Not really. Generic driver does support showing schemas in the navigator tree. Take a look at Clickhouse's meta-model.

**Reply by yangmeng66** [reply]
What exactly do you mean? This is "plugin. XML" of Clickhouse
How to make him show the folder "schemas"

```
<plugin>

    <extension point="org.jkiss.dbeaver.generic.meta">
        <meta id="clickhouse" class="org.jkiss.dbeaver.ext.clickhouse.model.ClickhouseMetaModel" driverClass="ru.yandex.clickhouse.ClickHouseDriver" dialect="clickhouse"/>
    </extension>

    
</plugin>
```

**Reply by LonwoLonwo** [reply]
```
<folder type="org.jkiss.dbeaver.ext.oracle.model.OracleSchema" label="%tree.schemas.node.name" icon="#folder_schema" description="Database schemas">
                    <items label="%tree.schema.node.name" path="schema" property="schemas" icon="#schema">
```

You can use folder annotation to see the Schemas folder in your tree.

**Reply by yangmeng66** [reply]
Hello, I understand this custom folder. My question is, can I put the default database "default and system" into my custom folder? Because I find it complicated to re customize everything, just like oracle

**Reply by LonwoLonwo** [reply]
Example of this request you can find here: https://github.com/dbeaver/dbeaver/blob/devel/plugins/org.jkiss.dbeaver.ext.hana/src/org/jkiss/dbeaver/ext/hana/model/HANAMetaModel.java#L73
(about PUBLIC scheme)

**Reply by yangmeng66** [reply]
hello @ShadelessFox ,
Where is the execution location of the SQL selected in the dbeaver UI in the code?

![image](https://user-images.githubusercontent.com/51857567/149432868-c0c9369d-0ba1-46be-8385-dbaa3f591f62.png)

**Reply by yangmeng66** [reply]
@ShadelessFox
  Hello, can you show me how to add a new project to the existing driver plug-in? It doesn't seem compatible
![image](https://user-images.githubusercontent.com/51857567/149490280-e9870496-2f92-42d5-9499-cec2479e6d5b.png)

**Reply by ShadelessFox** [reply]
You need to enable the desired plug-in for the product.

Go go <kbd>Debug Configurations</kbd> &rArr; <kbd>DBeaver.product</kbd> &rArr; <kbd>Plug-ins</kbd> and enable it here.

**Reply by yangmeng66** [reply]
Thank you very much.

**Reply by yangmeng66** [reply]
hello, @ShadelessFox 
Do you know how to print out all SQL statements interacting with the database?

**Reply by ShadelessFox** [reply]
You can view executed queries in Query Manager. See https://github.com/dbeaver/dbeaver/wiki/Query-Manager

**Comment by yangmeng66** [comment]
No, no, no, I don't mean that. What I want is the SQL driven by dbeaver to interact with the corresponding database, such as Oracle to obtain all tablespaces:
![image](https://user-images.githubusercontent.com/51857567/149732240-7dc8ec82-0213-482e-b3a4-4bd644bd980e.png)

**Reply by ShadelessFox** [reply]
They are shown too in Query Manager. Otherwise, I'm not sure what your question is.

**Reply by yangmeng66** [reply]
![image](https://user-images.githubusercontent.com/51857567/149741515-87741ebf-e56d-4e52-825c-ed52754e2a91.png)

**Reply by ShadelessFox** [reply]
Metadata queries are hidden by default. You need to enable it:
![image](https://user-images.githubusercontent.com/35821147/149743022-d72f383c-c56f-498d-92bc-cb816a457822.png)

**Reply by yangmeng66** [reply]
excellent, But why are so many empty?
![image](https://user-images.githubusercontent.com/51857567/149744042-6de7ed68-1bb3-4a36-948e-17b82abf964c.png)

**Reply by ShadelessFox** [reply]
These are generic ones. They use JDBC for retrieving metadata rather than performing queries themselves.

**Reply by yangmeng66** [reply]
Thank you very much and good luck

**Reply by yangmeng66** [reply]
Hello, @ShadelessFox 
Is this addfunctions useful? I found that deleting this line of code can also execute these functions. Why?

```
public class ClickhouseSQLDialect extends GenericSQLDialect {

    private static final String[] CLICKHOUSE_FUNCTIONS = {
        "quantile",
        "quantileExact",
        "uniq",
        "concat",
        "replaceOne",
         ......
}

  public void initDriverSettings(JDBCSession session, JDBCDataSource dataSource, JDBCDatabaseMetaData metaData) {
        super.initDriverSettings(session, dataSource, metaData);
        removeSQLKeyword("DEFAULT");  //What does this comment keyword mean?
        removeSQLKeyword("SYSTEM");   //What does this comment keyword mean?
        addFunctions(Arrays.asList(CLICKHOUSE_FUNCTIONS)); //You can also execute these functions by deleting this line of code
    }
```

**Reply by ShadelessFox** [reply]
It is only used for highlighting purposes. It does not affect execution.

**Reply by yangmeng66** [reply]
1. OK, that is, clickhousesqldialect is optional
2. In fact, I've been wondering what's the difference between the three
![image](https://user-images.githubusercontent.com/51857567/149857248-76c29b86-210a-496e-a655-3965285b80c1.png)

**Reply by ShadelessFox** [reply]
As of now, it is used for displaying these configuration buttons for filters:

![image](https://user-images.githubusercontent.com/35821147/149904999-3ca24acd-63e5-4a0e-abb1-a50711a85a76.png)

**Reply by yangmeng66** [reply]
well.
<1>  If I want to replace SQL1 in dbeaver UI with other SQL2, what should I do in the code?
<2>  Can you tell me what I should think of Oracle Is this the UI module? I can't do it now
<3> What is the specific function of this "<extension point="org.jkiss.dbeaver.objectManager">"?
 
`<manager class="org.jkiss.dbeaver.ext.oracle.edit.OracleTableManager" objectType="org.jkiss.dbeaver.ext.oracle.model.OracleTable"/>`

**Reply by ShadelessFox** [reply]
I do not quite understand the first two questions. For the latter one see https://github.com/dbeaver/dbeaver/discussions/15125#discussioncomment-1938639

**Reply by yangmeng66** [reply]
Oh, thank you
